### DISCUSSION TITLE
PDF changing on open and save?

### DISCUSSION THREAD

**Discussion by oconnor0** [empira/PDFsharp#184] [Q&A] [answered]
Hi,

I'm looking to modify some existing PDFs I have but wanted to make certain that I could open and save them without loss. I created the following .NET project:

```csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="PdfSharp" Version="6.1.1" />
  </ItemGroup>
</Project>
```

And the following Program.cs:

```csharp
using PdfSharp;
using PdfSharp.Drawing;
using PdfSharp.Pdf;
using PdfSharp.Pdf.IO;

PdfDocument input = PdfReader.Open(args[0], PdfDocumentOpenMode.Modify);
PdfDocument output = input;
output.Save(args[1]);
```

When I run the following:

```shell
$ dotnet run -- Input.pdf Output.pdf
$ ls -l Input.pdf Output.pdf
-rw-r--r-- 1 oco oco 31070533 Jun 14  2023 Input.pdf
-rw-rw-r-- 1 oco oco 30917970 Oct 12 21:28 Output.pdf
```

I get an output PDF that is slightly smaller than the input PDF. Obviously they are not identical so my questions are:

1. Why is this? Is there a way to get an identical output?
2. If not, is there a way to compare every aspect the PDFs?

**Comment by ThomasHoevel** [comment, marked as answer]
1. Currently there is no way to prevent this with PDFsharp. With support for "Incremental Updates" there will be a way to reduce the differences.
2. There is no easy way.

PDFsharp reads everything into memory and writes a completely new file on saving. Some orphaned objects may be dropped in that process.
Parts of PDF files are zipped. This is a lossless process, but size can change if an object is unzipped and re-zipped.
PDFsharp does not support XREF streams, but that usually leads to larger PDF files after saving.

It is normal that the file size changes.

**Reply by oconnor0** [reply]
Yeah, that sounds reasonable. It sounds like this means that the PDF I generated in here is equivalent but not identical.

Will I need to worry about silent loss of information here on unsupported features or does PDFsharp throw in the face of those?
